### PR TITLE
New device discovery

### PR DESCRIPTION
**Pull request by fsuc**

### CODE REVIEW

File: OCast/Core/DeviceDiscovery/SSDPMSearchRequest.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /// The max time header value.
+    let maxTime: Int
+    
+    /// The search target header value.
+    let searchTarget: String
+    
+    /// The data which represents a M-SEARCH request given structure properties.
+    var data: Data? {
+        
+        var requestString = requestLine + "\r\n"
```
Comment (chazemar) [inline OCast/Core/DeviceDiscovery/SSDPMSearchRequest.swift]: maybe use multiline string literals instead append()

File: OCast/Core/DeviceDiscovery/DeviceDiscovery.swift [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+/// Protocol for responding to device discovery events.
+@objc public protocol DeviceDiscoveryDelegate: class {
+    
+    /// Tells the delegate that new devices are found.
+    ///
+    /// - Parameters:
+    ///   - deviceDiscovery: The device discovery informing the delegate.
+    ///   - devices: The new devices found.
+    func deviceDiscovery(_ deviceDiscovery: DeviceDiscovery, didAddDevices devices: [Device])
```
Comment (chazemar) [inline OCast/Core/DeviceDiscovery/DeviceDiscovery.swift]: replace didAddDevices by didAdd

File: OCast/Core/DeviceDiscovery/DeviceDiscovery.swift [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    
+    /// Tells the delegate that devices are lost.
+    ///
+    /// - Parameters:
+    ///   - deviceDiscovery: The device discovery informing the delegate.
+    ///   - devices: The devices lost.
+    func deviceDiscovery(_ deviceDiscovery: DeviceDiscovery, didRemoveDevices devices: [Device])
```
Comment (chazemar) [inline OCast/Core/DeviceDiscovery/DeviceDiscovery.swift]: replace didRemoveDevices by didRemove

File: OCast/Core/DeviceDiscovery/DeviceDiscovery.swift [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    
+    /// Tells the delegate that the discovery is stopped. All the devices are removed.
+    ///
+    /// - Parameters:
+    ///   - deviceDiscovery: The device discovery informing the delegate.
+    ///   - error: The error if there's an issue, `nil` if the device discovery has been stopped normally.
+    func deviceDiscoveryDidStop(_ deviceDiscovery: DeviceDiscovery, withError error: Error?)
```
Comment (chazemar) [inline OCast/Core/DeviceDiscovery/DeviceDiscovery.swift]: replace withError by with

File: OCast/Core/DeviceDiscovery/DeviceDiscovery.swift [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    private var removeDevicesTasks = [DispatchWorkItem]()
+    
+    /// `true` if the discovery is paused, otherwise `false`
+    private var paused = false
+    
+    /// `true` if the discovery is stopped, otherwise `false`
+    private var stopped = true
+    
+    /// `true` if the discovery is running, otherwise `false`
+    private var isRunning: Bool {
```
Comment (chazemar) [inline OCast/Core/DeviceDiscovery/DeviceDiscovery.swift]: replace isRunning by running (or replace paused/stopped by isPaused/isStopped)

File: OCast/Core/DeviceDiscovery/UPNPService.swift [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                      friendlyName: friendlyName,
+                      manufacturer: manufacturer,
+                      modelName: modelName)
+    }
+    
+    /// Returns the application URL from an HTTP response.
+    ///
+    /// - Parameter httpHeaders: The HTTP response headers.
+    /// - Returns: The application URL or `nil`if an error occurs.
+    private func applicationURL(from httpHeaders: [AnyHashable: Any]) -> String? {
```
Comment (chazemar) [inline OCast/Core/DeviceDiscovery/UPNPService.swift]: replace from by fromHttpHeaders

File: OCast/Core/DeviceDiscovery/DeviceDiscovery.swift [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        return !stopped && !paused
+    }
+    
+    // The delegate to receive the device discovery events.
+    public weak var delegate: DeviceDiscoveryDelegate?
+    
+    /// The devices discovered on the network
+    public var devices: [Device] {
+        return Array(discoveredDevices.values)
```
[the remark this answers is not in the document]
Comment (fsuc) [reply OCast/Core/DeviceDiscovery/DeviceDiscovery.swift]: Yes, it's better than to keep another array just for that.

**Pull request merged**